OneCryptoPkg/OneCryptoLoaders/OneCryptoLoaderDxe.c

```diff
@@ -0,0 +1,489 @@
+/** @file
+  OneCryptoLoaderDxe.c
+
+  Copyright (c) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+  This file contains the implementation of the OneCryptoLoader DXE driver,
+  which is responsible for loading and initializing the shared cryptographic
+  library and its dependencies.
+
+**/ 
+
+#include <Library/DebugLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/SafeIntLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/DxeServicesLib.h>
+#include <Library/DebugLib.h>
+#include <Library/RngLib.h>
+#include <Library/PeCoffGetEntryPointLib.h>
+#include <Protocol/Rng.h>
+#include <Protocol/OneCrypto.h>
+
+#include <Private/OneCryptoDependencySupport.h>
+#include "OneCryptoLoaderDriverSupport.h"
+#include "PeCoffLib.h"
+
+#define EFI_SECTION_PE32  0x10
+
+//
+// Exposing gDriverDependencies to reduce phase specific code
+//
+DRIVER_DEPENDENCIES  *gDriverDependencies = NULL;
+//
+// The dependencies of the shared library, must live as long
+// as the shared code is used
+//
+ONE_CRYPTO_DEPENDENCIES  *mOneCryptoDepends = NULL;
+//
+// Crypto protocol for the shared library
+// Using VOID* to be agnostic about protocol structure size/layout
+//
+VOID  *mOneCryptoProtocol = NULL;
+
+//
+// Lazy RNG state tracking
+//
+STATIC BOOLEAN mRngInitAttempted = FALSE;
+STATIC EFI_RNG_PROTOCOL *mCachedRngProtocol = NULL;
+
+/**
+ * @brief Asserts that the given EFI_STATUS is not an error.
+ *
+ * This macro checks if the provided EFI_STATUS value indicates an error.
+ * If the status is an error, it triggers an assertion failure.
+ *
+ * @param Status The EFI_STATUS value to be checked.
+ */
+VOID
+EFIAPI
+AssertEfiError (
+  BOOLEAN  Expression
+  )
+{
+  ASSERT_EFI_ERROR (Expression);
+}
+
+/**
+ * @brief Lazy RNG implementation that locates EFI_RNG_PROTOCOL on first use
+ *
+ * This function implements lazy initialization of the RNG protocol to avoid
+ * boot-time hangs. It only attempts to locate the protocol when RNG is first
+ * needed, and caches the result for subsequent calls.
+ *
+ * @param[out] Rand Pointer to buffer to receive the 64-bit random number
+ * @return TRUE if random number generated successfully, FALSE otherwise
+ */
+BOOLEAN
+EFIAPI
+LazyPlatformGetRandomNumber64 (
+  OUT UINT64  *Rand
+  )
+{
+  EFI_STATUS  Status;
+  UINT8       *RandBytes;
+
+  if (Rand == NULL) {
+    DEBUG ((DEBUG_ERROR, "LazyPlatformGetRandomNumber64: Null Rand pointer\n"));
+    return FALSE;
+  }
+
+  //
+  // Only attempt to locate the RNG protocol once
+  //
+  if (!mRngInitAttempted) {
```

Comment:
Seems like it would be more useful to keep trying to get the RNG protocol on subsequent times if you didn't find it the first time